lib/graphql/schema/timeout_middleware.rb

```diff
@@ -0,0 +1,63 @@
+module GraphQL
+  class Schema
+    # This middleware will stop resolving new fields after `max_seconds` have elapsed.
+    # After the time has passed, any remaining fields will be `nil`, with errors added
+    # to the `errors` key. Any already-resolved fields will be in the `data` key, so
+    # you'll get a partial response.
+    #
+    # Note that this will stop a query _in between_ field resolutions, but
+    # it doesn't interrupt long-running `resolve` functions. Be sure to use
+    # timeout options for external connections. For more info, see
+    # www.mikeperham.com/2015/05/08/timeout-rubys-most-dangerous-api/
+    #
+    # @example Stop resolving fields after 2 seconds
+    #   MySchema.middleware << GraphQL::Schema::TimeoutMiddleware.new(max_seconds: 2)
+    #
+    class TimeoutMiddleware
+      # This key is used for storing timeout data in the {Query::Context} instance
+      DEFAULT_CONTEXT_KEY = :__timeout_at__
+      # @param max_seconds [Numeric] how many seconds the query should be allowed to resolve new fields
+      # @param context_key [Symbol] what key should be used to read and write to the query context
+      def initialize(max_seconds:, context_key: DEFAULT_CONTEXT_KEY)
```

Comment:
As it is, a user _must_ pass a value, otherwise :boom: :

```
irb(main):001:0> GraphQL::Schema::TimeoutMiddleware.new
ArgumentError: missing keyword: max_seconds
	from /Users/rmosolgo/projects/graphql/lib/graphql/schema/timeout_middleware.rb:21:in `initialize'
```

Is that the case you're concerned about here? 

Of course, it's still possible to pass nil as 

```ruby 
GraphQL::Schema::TimeoutMiddleware.new(max_seconds: nil)
```

But I'm not sure of a method signature that would prevent passing `nil`. If you think it's important to handle this kind of usage, we could add runtime check!